File: algo_dart/linked_list.go
package main

import "fmt"

type Node[E any] struct {
	Value E
	Next  *Node[E]
}

func (n *Node[E]) String() string {
	if n.Next == nil {
		return fmt.Sprintf("%v", n.Value)
	}
	return fmt.Sprintf("%v -> %v", n.Value, n.Next.String())
}

type LinkedList[E any] struct {
	head *Node[E]
	tail *Node[E]
}

func (list *LinkedList[E]) IsEmpty() bool {
	return list.head == nil
}

func (list *LinkedList[E]) Push(value E) {
	head := &Node[E]{
		Value: value,
		Next:  list.head,
	}
	if list.tail == nil {
		list.tail = head
	}
}

func (list *LinkedList[E]) Append(value E) {
	if list.IsEmpty() {
		list.Push(value)
		return
	}
	list.tail.Next = &Node[E]{
		Value: value,
	}

	list.tail = list.tail.Next

}

func (list *LinkedList[E]) NodeAt(index int) *Node[E] {
	currentIndex := 0
	currentNode := list.head

	for currentNode != nil && currentIndex != index {
		currentNode = currentNode.Next
		currentIndex++
	}

	return currentNode
}

func (list *LinkedList[E]) InsertAfter(node Node[E], value E) *Node[E] {
	if list.tail == &node {
		list.Append(value)
		return list.tail
	}

	node.Next = &Node[E]{Value: value, Next: node.Next}
	return node.Next
}

func (list *LinkedList[E]) Pop() E {
	value := list.head
	list.head = list.head.Next
	if list.IsEmpty() {
		list.tail = nil
	}
	return value.Value
}

func (list *LinkedList[E]) RemoveLast() E {
	if list.head.Next == nil {
		return list.Pop()
	}
	current := list.head

	for current != list.tail {
		current = current.Next
	}
	value := list.tail.Value
	list.tail = current
	current.Next = nil
	return value

}

func (list *LinkedList[E]) RemoveAfter(node *Node[E]) E {
	value := node.Next.Value
	if node.Next == list.tail {
		list.tail = node
	}
	node.Next = node.Next.Next

	return value
}

File: algo_dart/linked_list.dart
class Node<T> {
  T value;
  Node<T>? next;

  Node({required this.value, this.next});

  @override
  String toString() {
    if (next == null) return "$value";
    return "$value -> ${next.toString()}";
  }
}

void main() {
  // final node1 = Node(value: 1);
  // final node2 = Node(value: 2);
  // final node3 = Node(value: 3);

  // node1.next = node2;
  // node2.next = node3;
  // print(node1);

  final list = LinkedList<int>();
  list.push(1);
  list.append(2);
  list.append(3);
  list.append(4);
  print(list);
}

class LinkedList<E> extends Iterable<E> {
  Node<E>? head;
  Node<E>? tail;

  @override
  bool get isEmpty => head == null;

  @override
  String toString() {
    if (isEmpty) return 'Empty list';
    return head.toString();
  }

  void push(E value) {
    head = Node<E>(value: value, next: head);
    tail ??= head;
  }

  void append(E value) {
    if (isEmpty) {
      push(value);
      return;
    }

    tail?.next = Node(value: value);
    tail = tail?.next;
  }

  Node<E>? nodeAt(int index) {
    var currentNode = head;
    var currentIndex = 0;
    while (currentNode != null && currentIndex < index) {
      currentNode = currentNode.next;
      currentIndex++;
    }
    return currentNode;
  }

  Node<E> insertAfter(Node<E> node, E value) {
    if (tail == node) {
      append(value);
      return tail!;
    }

    node.next = Node(value: value, next: node.next);
    return node.next!;
  }

  E? pop() {
    final value = head;

    head = head?.next;
    if (isEmpty) {
      tail = null;
    }
    return value?.value;
  }

  E? removeLast() {
    if (head?.next == null) return pop();
    var current = head;
    while (current?.next != tail) {
      current = current?.next;
    }
    final value = tail?.value;
    tail = current;
    tail?.next = null;
    return value;
  }

  E? removeAfter(Node<E> node) {
    final value = node.next?.value;
    if (node.next == tail) {
      tail = node;
    }
    node.next = node.next?.next;
    return value;
  }

  @override
  Iterator<E> get iterator => _LinkedListIterator(this);
}

class _LinkedListIterator<E> implements Iterator<E> {
  _LinkedListIterator(LinkedList<E> list) : _list = list;
  final LinkedList<E> _list;

  Node<E>? _currentNode;

  @override
  E get current => _currentNode!.value;

  bool _firstPass = true;

  @override
  bool moveNext() {
    if (_list.isEmpty) return false;
    if (_firstPass) {
      _currentNode = _list.head;
      _firstPass = false;
    } else {
      _currentNode = _currentNode?.next;
    }

    return _currentNode != null;
  }
}

extension ReversibleLinkedList<E> on LinkedList<E> {
  void reverse() {
    tail = head;
    var previous = head;
    var current = head?.next;
    previous?.next = null;
    while (current != null) {
      final next = current.next;
      current.next = previous;
      previous = current;
      current = next;
    }
  }

  void reverse2() {
    tail = head;
    var current = head?.next;
    var previous = head;
    previous?.next = null;
    while (current != null) {
      final next = current.next;
      current.next = previous;
      previous = current;
      current = next;
    }
  }

  void removeAll(E value) {
    while (head != null && head?.value == value) {
      head = head?.next;
    }
    var previous = head;
    var current = head?.next;

    while (current != null) {
      if (current.value == value) {
        previous?.next = current.next;
        current = previous?.next;
        continue;
      }
      previous = current;
      current = current.next;
    }
    if (tail?.value == value) {
      tail = previous;
    }
  }
}

void printInReverse(LinkedList<int> list) {
  if (list.isEmpty) {
    print('Empty List');
  } else {
    nodePrinter(list.head);
  }
}

void nodePrinter(Node<int>? node) {
  if (node == null) {
    return;
  }
  nodePrinter(node.next);
  print(node.value);
}

Node<int>? findTheMiddleNode(LinkedList<int> list) {
  var slow = list.head;
  var fast = list.head;

  while (fast?.next != null) {
    fast = fast?.next?.next;
    slow = slow?.next;
  }

  return slow;
}

LinkedList<int> reverseALinkedList(LinkedList<int> list) {
  final reversedList = LinkedList<int>();
  var currentNode = list.head;
  while (currentNode != null) {
    reversedList.push(currentNode.value);
    currentNode = currentNode.next;
  }
  return reversedList;
}
